src/d3a/models/strategy/home_meter.py

```diff
@@ -0,0 +1,577 @@
+        self.bid_update = TemplateStrategyBidUpdater(
+            initial_rate=self.initial_buying_rate,
+            final_rate=self.final_buying_rate,
+            fit_to_limit=self.fit_to_limit,
+            energy_rate_change_per_update=self.energy_rate_increase_per_update,
+            update_interval=self.update_interval,
+            rate_limit_object=min)
+
+        self.offer_update = TemplateStrategyOfferUpdater(
+            initial_rate=self.initial_selling_rate,
+            final_rate=self.final_selling_rate,
+            fit_to_limit=self.fit_to_limit,
+            energy_rate_change_per_update=self.energy_rate_decrease_per_update,
+            update_interval=self.update_interval,
+            rate_limit_object=max)
+
+    def event_activate(self, **kwargs):
+        """Activate the device."""
+        self.event_activate_price()
+        self.event_activate_energy()
+        self.bid_update.update_and_populate_price_settings(self.area)
+        self.offer_update.update_and_populate_price_settings(self.area)
+
+    def event_activate_price(self):
+        """Configure all the rates for the device (both consumption and production)."""
+        # If we want to use the Market Maker rate, we must overwrite the existing rates with it.
+        if self.use_market_maker_rate:
+            self._replace_rates_with_market_maker_rates()
+
+        self._validate_consumption_rates(
+            initial_rate=self.bid_update.initial_rate_profile_buffer,
+            final_rate=self.bid_update.final_rate_profile_buffer,
+            energy_rate_change_per_update=(
+                self.bid_update.energy_rate_change_per_update_profile_buffer),
+            fit_to_limit=self.bid_update.fit_to_limit)
+
+        self._validate_production_rates(
+            initial_rate=self.offer_update.initial_rate_profile_buffer,
+            final_rate=self.offer_update.final_rate_profile_buffer,
+            energy_rate_change_per_update=(
+                self.offer_update.energy_rate_change_per_update_profile_buffer),
+            fit_to_limit=self.offer_update.fit_to_limit)
+
+    def event_activate_energy(self):
+        """Read the power profile and update the energy requirements for future market slots.
+
+        This method is triggered by the ACTIVATE event.
+        """
+        self.profile = self._read_raw_profile_data(self.home_meter_profile)
+        self._simulation_start_timestamp = self.area.now
+        self._set_energy_forecast_for_future_markets(reconfigure=True)
+
+    def event_market_cycle(self):
+        """Prepare rates and execute bids/offers when a new market slot begins.
+
+        This method is triggered by the MARKET_CYCLE event.
+        """
+        super().event_market_cycle()
+
+        self._reset_rates_and_update_prices()
+        self._set_energy_forecast_for_future_markets(reconfigure=False)
+
+        # TODO: should we always do both bids and offers, or should we first check the +/- of the
+        #  profile?
```

Comment:
These approaches are more or less equal (since no bid/offerwill be placed if the profile is negative/positive, the check is just not here but in the _post_offer and _post_first_bid respectively). If you want to be more explicit and not call one of the 2 methods be my guest, however I would leave it as it is. After all, you do not gain much in performance (market cycles are relatively infrequent)

Reply:
Thanks. I'll replace the TODO comment then 👍